### PR TITLE
Rebuild study spaces

### LINKED ISSUE #89: Rebuild Study Space API
The API should have four components:

1. Fetch study space location IDs
2. Fetch room availability for a given space ID and date
3. Book room for a given set of room IDs

Sample JSON for (1):
```json
{ 
    "locations": [
        {
            "id": 512,
            "name": "Education Commons",
            "service": "libcal"
        },
        {
            "id": 26,
            "name": "Weigle",
            "service": "libcal"
        },
        {
            ...
        }
    ]
}
```
Sample JSON for (2):
```json
{
    "location_id": 512,
    "date": "2018-01-25",
    "rooms": [
        {
            "room_id": 3,
            "name": "VP WIC Booth 01 (Capacity 6)",
            "image_url": "<URL HERE>",
            "times": [
                {
                    "booked": true,
                    "start": "5:30pm",
                    "end": "6:00pm"
                },
                {
                    "booked": false,
                    "start": "6:00pm",
                    "end": "6:30pm"
                },
                ...
            ]
        }, 
        {
        ...
        },
        {
        ...
        }    
    ]
}
```
Sample Request JSON during POST request for (3):
```json
{
           "location_id": 512,
           "room_id": 3,
           "times": [1, 2]
}
```

Sample Response JSON for (3) (until we can book the room ourself):
```json
{
           "success": true,
           "booking_url": "http://libcal.library.upenn.edu/reserve/spaces/vpdlc/"
}
```

### PR DESCRIPTION
**Pull request by ezwang**
Fix #89, rebuild the study space API. Implements everything needed to list rooms, but currently does not have any methods for booking rooms.

Tag to version `1.6.7` once this is merged.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -60,3 +60,5 @@ MANIFEST
 docs/_build
 .pypirc
 .idea/
+*.swp
+*.swo
```
Comment (adelq) [inline .gitignore]: Add to your global .gitignore

File: penn/studyspaces.py [hunk 2]
```diff
@@ -9,85 +15,76 @@ class StudySpaces(object):
...
+        return pytz.timezone("US/Eastern").localize(date)
 
-        roomTimes = []
+    @staticmethod
+    def get_room_id_name_mapping(building):
+        data = requests.get("{}/spaces?lid={}".format(BASE_URL, building)).content.decode("utf8")
+        # find all of the javascript room definitions
+        out = {}
+        for item in re.findall(r"resources.push\(((?s).*?)\);", data, re.MULTILINE):
+            items = {k: v for k, v in re.findall(r'(\w+?):\s*(.*?),', item)}
```
Comment (adelq) [inline penn/studyspaces.py]: Could you instead use the Python JSON parsing once you've found the data by regex in the previous line?

File: requirements.txt [hunk 3]
```diff
@@ -1,6 +1,8 @@
...
-requests==2.4.3
-beautifulsoup4==4.4.1
+beautifulsoup4==4.6.0
 html5lib==0.999
 mock==2.0.0
 nameparser==0.4.0
+nose==1.3.7
+pytz==2017.3
+requests==2.4.3
+six==1.11.0
```
Comment (adelq) [inline requirements.txt]: Add flake8?

File: penn/studyspaces.py (hunk 2 above)
Comment (ezwang) [reply penn/studyspaces.py]: Tried that, the issue is that the keys are not enclosed in quotes, which Python thinks is invalid JSON syntax. I figured that parsing the values directly would be better than enclosing the keys in quotes using regex and then using Python JSON parsing.

**Pull request merged**